### PR TITLE
[DNM]Read Thread and Data sharing

### PR DESCRIPTION
**Pull request by JinheLin**
### What problem does this PR solve?

Issue Number: close #xxx

Problem Summary:

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
None
```

### CODE REVIEW

File: dbms/src/Storages/DeltaMerge/ReadThread/CPU.cpp [hunk 1]
```diff
@@ -0,0 +1,68 @@
+#include <Storages/DeltaMerge/ReadThread/CPU.h>
+#include <Poco/File.h>
+#include <Poco/DirectoryIterator.h>
+#include <string>
+
+namespace DB::DM
+{
+// In Linux a numa node is represented by a device directory, such as '/sys/devices/system/node/node0', ''/sys/devices/system/node/node01'.
```
Comment (flowbehappy) [inline dbms/src/Storages/DeltaMerge/ReadThread/CPU.cpp:8]: I wonder whether it is a common case on all platforms, other than linux. We better do some protection, like rollback strategy on other platforms.

File: dbms/src/Storages/DeltaMerge/File/DMFileReader.cpp [hunk 2]
```diff
@@ -25,6 +25,11 @@
 #include <Storages/Page/PageUtil.h>
 #include <fmt/format.h>
 
+#include <atomic>
+
+#include "Storages/DeltaMerge/File/DMFile.h"
+#include "common/logger_useful.h"
```
Comment (flowbehappy) [inline dbms/src/Storages/DeltaMerge/File/DMFileReader.cpp:31]:
```suggestion
#include <Storages/DeltaMerge/File/DMFile.h>
#include <common/logger_useful.h>
```

File: dbms/src/Storages/DeltaMerge/ReadThread/SegmentReadTaskScheduler.h [hunk 3]
```diff
@@ -0,0 +1,244 @@
...
+
+class DMFileReader;
+class DMFileReaderPool
+{
+public:
+    static DMFileReaderPool & instance();
+
+    void add(DMFileReader & reader);
+    void del(DMFileReader & reader);
+    void set(DMFileReader & from_reader, int64_t col_id, size_t start, size_t count, ColumnPtr & col);
```
Comment (flowbehappy) [inline dbms/src/Storages/DeltaMerge/ReadThread/SegmentReadTaskScheduler.h:239]: Please add some doc about those methods, and explain how they work.

File: dbms/src/Storages/DeltaMerge/ReadThread/SegmentReadTaskScheduler.h [hunk 4]
```diff
@@ -0,0 +1,244 @@
...
+            finished[i] = 1;
+            finished_count++;
+        }
+    }
+    size_t finished_count;
+    std::vector<int8_t> finished;
+};
+using MergedTaskPtr = std::shared_ptr<MergedTask>;
+
+class SegmentReadTaskScheduler
```
Comment (flowbehappy) [inline dbms/src/Storages/DeltaMerge/ReadThread/SegmentReadTaskScheduler.h:191]: Add docs about the algorithm of the scheduler.

File: dbms/src/Storages/DeltaMerge/ReadThread/SegmentReadTaskScheduler.h [hunk 5]
```diff
@@ -0,0 +1,244 @@
...
+{
+public:
+    static DMFileReaderPool & instance();
+
+    void add(DMFileReader & reader);
+    void del(DMFileReader & reader);
+    void set(DMFileReader & from_reader, int64_t col_id, size_t start, size_t count, ColumnPtr & col);
+private:
+    std::mutex mtx;
+    std::unordered_map<uint64_t, std::unordered_set<DMFileReader*>> readers;
```
Comment (flowbehappy) [inline dbms/src/Storages/DeltaMerge/ReadThread/SegmentReadTaskScheduler.h:242]: I wonder whether it is safe to store raw pointer. Why not use smart pointer?

File: dbms/src/Storages/DeltaMerge/ReadThread/CPU.cpp (hunk 1 above)
Comment (SchrodingerZhu) [reply dbms/src/Storages/DeltaMerge/ReadThread/CPU.cpp:8]: some examples are at: https://github.com/microsoft/mimalloc/blob/master/src/os.c

File: dbms/src/Storages/DeltaMerge/ReadThread/CPU.cpp [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+    {
+        if (isCPU(iter.name()))
+        {
+            cpus.push_back(parseCPUNumber(iter.name()));
+        }
+    }
+    return cpus;
+}
+
+// TODO: What if the process running in the container and the CPU is limited.
```
Comment (SchrodingerZhu) [inline dbms/src/Storages/DeltaMerge/ReadThread/CPU.cpp:41]: I believe in a similar way we can get cpu and numa node configured for current PID, as it is stated in `libnuma`.

File: dbms/src/Storages/DeltaMerge/SegmentReadTaskPool.cpp [hunk 7]
```diff
@@ -95,4 +101,72 @@ SegmentReadTasks SegmentReadTask::trySplitReadTasks(const SegmentReadTasks & tas
...
+    }
+    LOG_FMT_DEBUG(log, "finishSegment pool {} seg_id {} pool_finished {}",
+        id, seg_id, pool_finished);
+    if (pool_finished)
+    {
+        q.finish();
+    }
+}
+
+SegmentReadTaskPtr SegmentReadTaskPool::getTask(uint64_t seg_id)
```
Comment (flowbehappy) [inline dbms/src/Storages/DeltaMerge/SegmentReadTaskPool.cpp:138]: `getTask` is not a good name, as we do lots of modifications in this method.

**Pull request closed without merging**